File: resources/android/src/com/nativephp/plugin/sensors/SensorsFunctions.kt
package com.nativephp.plugin.sensors

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import com.example.androidphp.bridge.BridgeFunction
import com.example.androidphp.bridge.BridgeResponse

object SensorsFunctions {

    // Cached sensor values
    private var accelerometerValues = floatArrayOf(0f, 0f, 0f)
    private var gyroscopeValues = floatArrayOf(0f, 0f, 0f)
    private var magnetometerValues = floatArrayOf(0f, 0f, 0f)
    private var rotationValues = floatArrayOf(0f, 0f, 0f, 0f, 0f)
    private var gravityValues = floatArrayOf(0f, 0f, 0f)
    private var pressureValue = 0f

    // Running state per sensor
    private var accelerometerRunning = false
    private var gyroscopeRunning = false
    private var magnetometerRunning = false
    private var deviceMotionRunning = false
    private var barometerRunning = false

    // Sensor availability
    private var hasAccelerometer = false
    private var hasGyroscope = false
    private var hasMagnetometer = false
    private var hasRotation = false
    private var hasGravity = false
    private var hasPressure = false

    private var sensorManager: SensorManager? = null
    private var initialized = false

    // Individual listeners for each sensor
    private val accelerometerListener = object : SensorEventListener {
        override fun onSensorChanged(event: SensorEvent) {
            accelerometerValues = event.values.copyOf()
        }
        override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}
    }

    private val gyroscopeListener = object : SensorEventListener {
        override fun onSensorChanged(event: SensorEvent) {
            gyroscopeValues = event.values.copyOf()
        }
        override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}
    }

    private val magnetometerListener = object : SensorEventListener {
        override fun onSensorChanged(event: SensorEvent) {
            magnetometerValues = event.values.copyOf()
        }
        override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}
    }

    private val rotationListener = object : SensorEventListener {
        override fun onSensorChanged(event: SensorEvent) {
            rotationValues = event.values.copyOf()
        }
        override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}
    }

    private val gravityListener = object : SensorEventListener {
        override fun onSensorChanged(event: SensorEvent) {
            gravityValues = event.values.copyOf()
        }
        override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}
    }

    private val pressureListener = object : SensorEventListener {
        override fun onSensorChanged(event: SensorEvent) {
            pressureValue = event.values[0]
        }
        override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}
    }

    private fun initialize(context: Context) {
        if (initialized) return

        sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
        val sm = sensorManager ?: return

        hasAccelerometer = sm.getDefaultSensor(Sensor.TYPE_ACCELEROMETER) != null
        hasGyroscope = sm.getDefaultSensor(Sensor.TYPE_GYROSCOPE) != null
        hasMagnetometer = sm.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD) != null
        hasRotation = sm.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR) != null
        hasGravity = sm.getDefaultSensor(Sensor.TYPE_GRAVITY) != null
        hasPressure = sm.getDefaultSensor(Sensor.TYPE_PRESSURE) != null

        initialized = true
    }

    // MARK: - Accelerometer

    class AccelerometerStart(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            initialize(context)

            if (accelerometerRunning) {
                return mapOf("success" to true, "message" to "Already running")
            }

            if (!hasAccelerometer) {
                return BridgeResponse.error("NOT_AVAILABLE", "Accelerometer not available")
            }

            val sm = sensorManager ?: return BridgeResponse.error("INIT_ERROR", "SensorManager not available")
            val sensor = sm.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
            sm.registerListener(accelerometerListener, sensor, SensorManager.SENSOR_DELAY_GAME)
            accelerometerRunning = true

            return mapOf("success" to true)
        }
    }

    class AccelerometerStop(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            initialize(context)

            if (!accelerometerRunning) {
                return mapOf("success" to true, "message" to "Not running")
            }

            sensorManager?.unregisterListener(accelerometerListener)
            accelerometerValues = floatArrayOf(0f, 0f, 0f)
            accelerometerRunning = false

            return mapOf("success" to true)
        }
    }

    class AccelerometerRead(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            initialize(context)

            if (!accelerometerRunning) {
                return BridgeResponse.error("NOT_STARTED", "Accelerometer not started. Call start() first.")
            }

            return mapOf(
                "x" to accelerometerValues[0].toDouble(),
                "y" to accelerometerValues[1].toDouble(),
                "z" to accelerometerValues[2].toDouble()
            )
        }
    }

    // MARK: - Gyroscope

    class GyroscopeStart(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            initialize(context)

            if (gyroscopeRunning) {
                return mapOf("success" to true, "message" to "Already running")
            }

            if (!hasGyroscope) {
                return BridgeResponse.error("NOT_AVAILABLE", "Gyroscope not available")
            }

            val sm = sensorManager ?: return BridgeResponse.error("INIT_ERROR", "SensorManager not available")
            val sensor = sm.getDefaultSensor(Sensor.TYPE_GYROSCOPE)
            sm.registerListener(gyroscopeListener, sensor, SensorManager.SENSOR_DELAY_GAME)
            gyroscopeRunning = true

            return mapOf("success" to true)
        }
    }

    class GyroscopeStop(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            initialize(context)

            if (!gyroscopeRunning) {
                return mapOf("success" to true, "message" to "Not running")
            }

            sensorManager?.unregisterListener(gyroscopeListener)
            gyroscopeValues = floatArrayOf(0f, 0f, 0f)
            gyroscopeRunning = false

            return mapOf("success" to true)
        }
    }

    class GyroscopeRead(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            initialize(context)

            if (!gyroscopeRunning) {
                return BridgeResponse.error("NOT_STARTED", "Gyroscope not started. Call start() first.")
            }

            return mapOf(
                "x" to gyroscopeValues[0].toDouble(),
                "y" to gyroscopeValues[1].toDouble(),
                "z" to gyroscopeValues[2].toDouble()
            )
        }
    }

    // MARK: - Magnetometer

    class MagnetometerStart(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            initialize(context)

            if (magnetometerRunning) {
                return mapOf("success" to true, "message" to "Already running")
            }

            if (!hasMagnetometer) {
                return BridgeResponse.error("NOT_AVAILABLE", "Magnetometer not available")
            }

            val sm = sensorManager ?: return BridgeResponse.error("INIT_ERROR", "SensorManager not available")
            val sensor = sm.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD)
            sm.registerListener(magnetometerListener, sensor, SensorManager.SENSOR_DELAY_GAME)
            magnetometerRunning = true

            return mapOf("success" to true)
        }
    }

    class MagnetometerStop(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            initialize(context)

            if (!magnetometerRunning) {
                return mapOf("success" to true, "message" to "Not running")
            }

            sensorManager?.unregisterListener(magnetometerListener)
            magnetometerValues = floatArrayOf(0f, 0f, 0f)
            magnetometerRunning = false

            return mapOf("success" to true)
        }
    }

    class MagnetometerRead(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            initialize(context)

            if (!magnetometerRunning) {
                return BridgeResponse.error("NOT_STARTED", "Magnetometer not started. Call start() first.")
            }

            return mapOf(
                "x" to magnetometerValues[0].toDouble(),
                "y" to magnetometerValues[1].toDouble(),
                "z" to magnetometerValues[2].toDouble()
            )
        }
    }

    // MARK: - Device Motion (rotation + gravity + gyroscope)

    class DeviceMotionStart(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            initialize(context)

            if (deviceMotionRunning) {
                return mapOf("success" to true, "message" to "Already running")
            }

            if (!hasRotation && !hasGravity && !hasGyroscope) {
                return BridgeResponse.error("NOT_AVAILABLE", "Device motion sensors not available")
            }

            val sm = sensorManager ?: return BridgeResponse.error("INIT_ERROR", "SensorManager not available")

            if (hasRotation) {
                val sensor = sm.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR)
                sm.registerListener(rotationListener, sensor, SensorManager.SENSOR_DELAY_GAME)
            }
            if (hasGravity) {
                val sensor = sm.getDefaultSensor(Sensor.TYPE_GRAVITY)
                sm.registerListener(gravityListener, sensor, SensorManager.SENSOR_DELAY_GAME)
            }
            if (hasGyroscope && !gyroscopeRunning) {
                // Only start gyroscope if not already running standalone
                val sensor = sm.getDefaultSensor(Sensor.TYPE_GYROSCOPE)
                sm.registerListener(gyroscopeListener, sensor, SensorManager.SENSOR_DELAY_GAME)
            }

            deviceMotionRunning = true
            return mapOf("success" to true)
        }
    }

    class DeviceMotionStop(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            initialize(context)

            if (!deviceMotionRunning) {
                return mapOf("success" to true, "message" to "Not running")
            }

            sensorManager?.unregisterListener(rotationListener)
            sensorManager?.unregisterListener(gravityListener)
            // Only stop gyroscope if not running standalone
            if (!gyroscopeRunning) {
                sensorManager?.unregisterListener(gyroscopeListener)
            }

            rotationValues = floatArrayOf(0f, 0f, 0f, 0f, 0f)
            gravityValues = floatArrayOf(0f, 0f, 0f)
            deviceMotionRunning = false

            return mapOf("success" to true)
        }
    }

    class DeviceMotionRead(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            initialize(context)

            if (!deviceMotionRunning) {
                return BridgeResponse.error("NOT_STARTED", "Device motion not started. Call start() first.")
            }

            // Convert rotation vector to euler angles
            val rotationMatrix = FloatArray(9)
            val orientation = FloatArray(3)

            if (hasRotation && rotationValues.size >= 4) {
                SensorManager.getRotationMatrixFromVector(rotationMatrix, rotationValues)
                SensorManager.getOrientation(rotationMatrix, orientation)
            }

            return mapOf(
                "attitude" to mapOf(
                    "pitch" to orientation[1].toDouble(),
                    "roll" to orientation[2].toDouble(),
                    "yaw" to orientation[0].toDouble()
                ),
                "rotationRate" to mapOf(
                    "x" to gyroscopeValues[0].toDouble(),
                    "y" to gyroscopeValues[1].toDouble(),
                    "z" to gyroscopeValues[2].toDouble()
                ),
                "gravity" to mapOf(
                    "x" to gravityValues[0].toDouble(),
                    "y" to gravityValues[1].toDouble(),
                    "z" to gravityValues[2].toDouble()
                )
            )
        }
    }

    // MARK: - Barometer

    class BarometerStart(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            initialize(context)

            if (barometerRunning) {
                return mapOf("success" to true, "message" to "Already running")
            }

            if (!hasPressure) {
                return BridgeResponse.error("NOT_AVAILABLE", "Barometer not available")
            }

            val sm = sensorManager ?: return BridgeResponse.error("INIT_ERROR", "SensorManager not available")
            val sensor = sm.getDefaultSensor(Sensor.TYPE_PRESSURE)
            sm.registerListener(pressureListener, sensor, SensorManager.SENSOR_DELAY_NORMAL)
            barometerRunning = true

            return mapOf("success" to true)
        }
    }

    class BarometerStop(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            initialize(context)

            if (!barometerRunning) {
                return mapOf("success" to true, "message" to "Not running")
            }

            sensorManager?.unregisterListener(pressureListener)
            pressureValue = 0f
            barometerRunning = false

            return mapOf("success" to true)
        }
    }

    class BarometerRead(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            initialize(context)

            if (!barometerRunning) {
                return BridgeResponse.error("NOT_STARTED", "Barometer not started. Call start() first.")
            }

            val altitude = SensorManager.getAltitude(SensorManager.PRESSURE_STANDARD_ATMOSPHERE, pressureValue)

            return mapOf(
                "pressure" to pressureValue.toDouble(),
                "altitude" to altitude.toDouble()
            )
        }
    }

    // MARK: - Availability

    class IsAvailable(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            initialize(context)

            return mapOf(
                "accelerometer" to hasAccelerometer,
                "gyroscope" to hasGyroscope,
                "magnetometer" to hasMagnetometer,
                "deviceMotion" to (hasRotation || hasGravity || hasGyroscope),
                "barometer" to hasPressure
            )
        }
    }
}

File: resources/ios/Sources/SensorsFunctions.swift
import Foundation
import CoreMotion

enum SensorsFunctions {

    // Lazy initialization to avoid crashes at app startup
    private static var _motionManager: CMMotionManager?
    private static var _altimeter: CMAltimeter?

    private static var motionManager: CMMotionManager {
        if _motionManager == nil {
            _motionManager = CMMotionManager()
        }
        return _motionManager!
    }

    private static var altimeter: CMAltimeter {
        if _altimeter == nil {
            _altimeter = CMAltimeter()
        }
        return _altimeter!
    }

    // Cached sensor values
    private static var accelerometerData: CMAccelerometerData?
    private static var gyroscopeData: CMGyroData?
    private static var magnetometerData: CMMagnetometerData?
    private static var deviceMotionData: CMDeviceMotion?
    private static var altimeterData: CMAltitudeData?

    // Running state per sensor
    private static var accelerometerRunning = false
    private static var gyroscopeRunning = false
    private static var magnetometerRunning = false
    private static var deviceMotionRunning = false
    private static var barometerRunning = false

    private static let updateInterval = 1.0 / 60.0 // 60 Hz

    // MARK: - Accelerometer

    class AccelerometerStart: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard !SensorsFunctions.accelerometerRunning else {
                return ["success": true, "message": "Already running"]
            }

            guard SensorsFunctions.motionManager.isAccelerometerAvailable else {
                return BridgeResponse.error(code: "NOT_AVAILABLE", message: "Accelerometer not available")
            }

            if !Thread.isMainThread {
                DispatchQueue.main.sync {
                    SensorsFunctions.motionManager.accelerometerUpdateInterval = SensorsFunctions.updateInterval
                    SensorsFunctions.motionManager.startAccelerometerUpdates(to: .main) { data, _ in
                        SensorsFunctions.accelerometerData = data
                    }
                }
            } else {
                SensorsFunctions.motionManager.accelerometerUpdateInterval = SensorsFunctions.updateInterval
                SensorsFunctions.motionManager.startAccelerometerUpdates(to: .main) { data, _ in
                    SensorsFunctions.accelerometerData = data
                }
            }

            SensorsFunctions.accelerometerRunning = true
            return ["success": true]
        }
    }

    class AccelerometerStop: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard SensorsFunctions.accelerometerRunning else {
                return ["success": true, "message": "Not running"]
            }

            if !Thread.isMainThread {
                DispatchQueue.main.sync {
                    SensorsFunctions.motionManager.stopAccelerometerUpdates()
                }
            } else {
                SensorsFunctions.motionManager.stopAccelerometerUpdates()
            }

            SensorsFunctions.accelerometerData = nil
            SensorsFunctions.accelerometerRunning = false
            return ["success": true]
        }
    }

    class AccelerometerRead: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard SensorsFunctions.accelerometerRunning else {
                return BridgeResponse.error(code: "NOT_STARTED", message: "Accelerometer not started. Call start() first.")
            }

            guard let data = SensorsFunctions.accelerometerData else {
                return ["x": 0.0, "y": 0.0, "z": 0.0]
            }

            // CoreMotion returns acceleration in g-force, convert to m/s²
            let gravity = 9.81
            return [
                "x": data.acceleration.x * gravity,
                "y": data.acceleration.y * gravity,
                "z": data.acceleration.z * gravity
            ]
        }
    }

    // MARK: - Gyroscope

    class GyroscopeStart: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard !SensorsFunctions.gyroscopeRunning else {
                return ["success": true, "message": "Already running"]
            }

            guard SensorsFunctions.motionManager.isGyroAvailable else {
                return BridgeResponse.error(code: "NOT_AVAILABLE", message: "Gyroscope not available")
            }

            if !Thread.isMainThread {
                DispatchQueue.main.sync {
                    SensorsFunctions.motionManager.gyroUpdateInterval = SensorsFunctions.updateInterval
                    SensorsFunctions.motionManager.startGyroUpdates(to: .main) { data, _ in
                        SensorsFunctions.gyroscopeData = data
                    }
                }
            } else {
                SensorsFunctions.motionManager.gyroUpdateInterval = SensorsFunctions.updateInterval
                SensorsFunctions.motionManager.startGyroUpdates(to: .main) { data, _ in
                    SensorsFunctions.gyroscopeData = data
                }
            }

            SensorsFunctions.gyroscopeRunning = true
            return ["success": true]
        }
    }

    class GyroscopeStop: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard SensorsFunctions.gyroscopeRunning else {
                return ["success": true, "message": "Not running"]
            }

            if !Thread.isMainThread {
                DispatchQueue.main.sync {
                    SensorsFunctions.motionManager.stopGyroUpdates()
                }
            } else {
                SensorsFunctions.motionManager.stopGyroUpdates()
            }

            SensorsFunctions.gyroscopeData = nil
            SensorsFunctions.gyroscopeRunning = false
            return ["success": true]
        }
    }

    class GyroscopeRead: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard SensorsFunctions.gyroscopeRunning else {
                return BridgeResponse.error(code: "NOT_STARTED", message: "Gyroscope not started. Call start() first.")
            }

            guard let data = SensorsFunctions.gyroscopeData else {
                return ["x": 0.0, "y": 0.0, "z": 0.0]
            }

            return [
                "x": data.rotationRate.x,
                "y": data.rotationRate.y,
                "z": data.rotationRate.z
            ]
        }
    }

    // MARK: - Magnetometer

    class MagnetometerStart: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard !SensorsFunctions.magnetometerRunning else {
                return ["success": true, "message": "Already running"]
            }

            guard SensorsFunctions.motionManager.isMagnetometerAvailable else {
                return BridgeResponse.error(code: "NOT_AVAILABLE", message: "Magnetometer not available")
            }

            if !Thread.isMainThread {
                DispatchQueue.main.sync {
                    SensorsFunctions.motionManager.magnetometerUpdateInterval = SensorsFunctions.updateInterval
                    SensorsFunctions.motionManager.startMagnetometerUpdates(to: .main) { data, _ in
                        SensorsFunctions.magnetometerData = data
                    }
                }
            } else {
                SensorsFunctions.motionManager.magnetometerUpdateInterval = SensorsFunctions.updateInterval
                SensorsFunctions.motionManager.startMagnetometerUpdates(to: .main) { data, _ in
                    SensorsFunctions.magnetometerData = data
                }
            }

            SensorsFunctions.magnetometerRunning = true
            return ["success": true]
        }
    }

    class MagnetometerStop: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard SensorsFunctions.magnetometerRunning else {
                return ["success": true, "message": "Not running"]
            }

            if !Thread.isMainThread {
                DispatchQueue.main.sync {
                    SensorsFunctions.motionManager.stopMagnetometerUpdates()
                }
            } else {
                SensorsFunctions.motionManager.stopMagnetometerUpdates()
            }

            SensorsFunctions.magnetometerData = nil
            SensorsFunctions.magnetometerRunning = false
            return ["success": true]
        }
    }

    class MagnetometerRead: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard SensorsFunctions.magnetometerRunning else {
                return BridgeResponse.error(code: "NOT_STARTED", message: "Magnetometer not started. Call start() first.")
            }

            guard let data = SensorsFunctions.magnetometerData else {
                return ["x": 0.0, "y": 0.0, "z": 0.0]
            }

            return [
                "x": data.magneticField.x,
                "y": data.magneticField.y,
                "z": data.magneticField.z
            ]
        }
    }

    // MARK: - Device Motion

    class DeviceMotionStart: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard !SensorsFunctions.deviceMotionRunning else {
                return ["success": true, "message": "Already running"]
            }

            guard SensorsFunctions.motionManager.isDeviceMotionAvailable else {
                return BridgeResponse.error(code: "NOT_AVAILABLE", message: "Device motion not available")
            }

            if !Thread.isMainThread {
                DispatchQueue.main.sync {
                    SensorsFunctions.motionManager.deviceMotionUpdateInterval = SensorsFunctions.updateInterval
                    SensorsFunctions.motionManager.startDeviceMotionUpdates(to: .main) { data, _ in
                        SensorsFunctions.deviceMotionData = data
                    }
                }
            } else {
                SensorsFunctions.motionManager.deviceMotionUpdateInterval = SensorsFunctions.updateInterval
                SensorsFunctions.motionManager.startDeviceMotionUpdates(to: .main) { data, _ in
                    SensorsFunctions.deviceMotionData = data
                }
            }

            SensorsFunctions.deviceMotionRunning = true
            return ["success": true]
        }
    }

    class DeviceMotionStop: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard SensorsFunctions.deviceMotionRunning else {
                return ["success": true, "message": "Not running"]
            }

            if !Thread.isMainThread {
                DispatchQueue.main.sync {
                    SensorsFunctions.motionManager.stopDeviceMotionUpdates()
                }
            } else {
                SensorsFunctions.motionManager.stopDeviceMotionUpdates()
            }

            SensorsFunctions.deviceMotionData = nil
            SensorsFunctions.deviceMotionRunning = false
            return ["success": true]
        }
    }

    class DeviceMotionRead: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard SensorsFunctions.deviceMotionRunning else {
                return BridgeResponse.error(code: "NOT_STARTED", message: "Device motion not started. Call start() first.")
            }

            guard let data = SensorsFunctions.deviceMotionData else {
                return [
                    "attitude": ["pitch": 0.0, "roll": 0.0, "yaw": 0.0],
                    "rotationRate": ["x": 0.0, "y": 0.0, "z": 0.0],
                    "gravity": ["x": 0.0, "y": 0.0, "z": 0.0]
                ]
            }

            return [
                "attitude": [
                    "pitch": data.attitude.pitch,
                    "roll": data.attitude.roll,
                    "yaw": data.attitude.yaw
                ],
                "rotationRate": [
                    "x": data.rotationRate.x,
                    "y": data.rotationRate.y,
                    "z": data.rotationRate.z
                ],
                "gravity": [
                    "x": data.gravity.x,
                    "y": data.gravity.y,
                    "z": data.gravity.z
                ]
            ]
        }
    }

    // MARK: - Barometer

    class BarometerStart: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard !SensorsFunctions.barometerRunning else {
                return ["success": true, "message": "Already running"]
            }

            guard CMAltimeter.isRelativeAltitudeAvailable() else {
                return BridgeResponse.error(code: "NOT_AVAILABLE", message: "Barometer not available")
            }

            if !Thread.isMainThread {
                DispatchQueue.main.sync {
                    SensorsFunctions.altimeter.startRelativeAltitudeUpdates(to: .main) { data, _ in
                        SensorsFunctions.altimeterData = data
                    }
                }
            } else {
                SensorsFunctions.altimeter.startRelativeAltitudeUpdates(to: .main) { data, _ in
                    SensorsFunctions.altimeterData = data
                }
            }

            SensorsFunctions.barometerRunning = true
            return ["success": true]
        }
    }

    class BarometerStop: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard SensorsFunctions.barometerRunning else {
                return ["success": true, "message": "Not running"]
            }

            if !Thread.isMainThread {
                DispatchQueue.main.sync {
                    SensorsFunctions.altimeter.stopRelativeAltitudeUpdates()
                }
            } else {
                SensorsFunctions.altimeter.stopRelativeAltitudeUpdates()
            }

            SensorsFunctions.altimeterData = nil
            SensorsFunctions.barometerRunning = false
            return ["success": true]
        }
    }

    class BarometerRead: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard SensorsFunctions.barometerRunning else {
                return BridgeResponse.error(code: "NOT_STARTED", message: "Barometer not started. Call start() first.")
            }

            guard let data = SensorsFunctions.altimeterData else {
                return ["pressure": 0.0, "altitude": 0.0]
            }

            return [
                "pressure": data.pressure.doubleValue * 10.0,
                "altitude": data.relativeAltitude.doubleValue
            ]
        }
    }

    // MARK: - Availability Check

    class IsAvailable: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            let mm = CMMotionManager()

            return [
                "accelerometer": mm.isAccelerometerAvailable,
                "gyroscope": mm.isGyroAvailable,
                "magnetometer": mm.isMagnetometerAvailable,
                "deviceMotion": mm.isDeviceMotionAvailable,
                "barometer": CMAltimeter.isRelativeAltitudeAvailable()
            ]
        }
    }
}
